src/main/java/org/gaul/s3proxy/EncryptedBlobStore.java

```diff
@@ -0,0 +1,784 @@
+    }
+
+    private boolean isEncrypted(String blobName) {
+        return blobName.endsWith(Constants.S3_ENC_SUFFIX);
+    }
+
+    private MutableBlobMetadata setEncryptedSuffix(BlobMetadata blobMeta) {
+        MutableBlobMetadata bm = new MutableBlobMetadataImpl(blobMeta);
+        if (blobMeta.getName() != null && !isEncrypted(blobMeta.getName())) {
+            bm.setName(blobNameWithSuffix(blobMeta.getName()));
+        }
+
+        return bm;
+    }
+
+    private String removeEncryptedSuffix(String blobName) {
+        return blobName.substring(0,
+            blobName.length() - Constants.S3_ENC_SUFFIX.length());
+    }
+
+    private MutableBlobMetadata removeEncryptedSuffix(BlobMetadata blobMeta) {
+        MutableBlobMetadata bm = new MutableBlobMetadataImpl(blobMeta);
+        if (isEncrypted(bm.getName())) {
+            String blobName = bm.getName();
+            bm.setName(removeEncryptedSuffix(blobName));
+        }
+
+        return bm;
+    }
+
+    private MutableBlobMetadata calculateBlobSize(BlobMetadata blobMeta) {
+        MutableBlobMetadata mbm = removeEncryptedSuffix(blobMeta);
+
+        // we are using on non-s3 backends like azure or gcp a metadata key to
+        // calculate the part padding sizes that needs to be removed
+        if (mbm.getUserMetadata()
+            .containsKey(Constants.METADATA_ENCRYPTION_PARTS)) {
+            int parts = Integer.parseInt(
+                mbm.getUserMetadata().get(Constants.METADATA_ENCRYPTION_PARTS));
+            int partPaddingSizes = Constants.PADDING_BLOCK_SIZE * parts;
+            long size = blobMeta.getSize() - partPaddingSizes;
+            mbm.setSize(size);
+            mbm.getContentMetadata().setContentLength(size);
+        } else {
+            // on s3 backends like aws or minio we rely on the eTag suffix
+            String regex = ".*-([0-9]+)";
+            Pattern pattern = Pattern.compile(regex);
+            Matcher matcher = pattern.matcher(blobMeta.getETag());
+            if (matcher.find()) {
+                int parts = Integer.parseInt(matcher.group(1));
+                int partPaddingSizes = Constants.PADDING_BLOCK_SIZE * parts;
+                long size = blobMeta.getSize() - partPaddingSizes;
+                mbm.setSize(size);
+                mbm.getContentMetadata().setContentLength(size);
+            } else {
+                long size = blobMeta.getSize() - Constants.PADDING_BLOCK_SIZE;
+                mbm.setSize(size);
+                mbm.getContentMetadata().setContentLength(size);
+            }
+        }
+
+        return mbm;
+    }
+
+    private boolean multipartRequiresStub() {
+        String blobStoreType = getBlobStoreType();
+        return Quirks.MULTIPART_REQUIRES_STUB.contains(blobStoreType);
+    }
+
+    private String blobNameWithSuffix(String container, String name) {
+        String nameWithSuffix = blobNameWithSuffix(name);
+        if (delegate().blobExists(container, nameWithSuffix)) {
```

Comment:
It would be better to avoid the HEAD request here. Can we always do a GET with the suffix and on 404 GET without it? I haven't internalized what changes would be required, but turning every GET into HEAD + GET is not desirable.

Reply:
since I tried to make it work in a mixed scenario where you can have encrypted and unencrypted files on the backend this is the trade of right now.

Reply:
Right, I understand the reasoning. I'm suggesting making an unconditional request for `<blob>.s3enc` and if it fails, making another request for `<blob>`, which in the case of encrypted blobs would result in a single request and in 2 requests for unencrypted blobs. Is it challenging to handle the error?

Reply:
Got your point, will look tomorrow if I can find a way doing it.

Reply:
I was reviewing the part and was only able to reduce one unnecessary HEAD call. 
As of now I don't see a possibility to get rid of the additional HEAD call therefore I extended the limitations in Encryption.md.
I hope that it is not a blocker from your side?

Reply:
I didn't notice the one more HEAD request -- thanks for fixing that up. I think we can table this for now and maybe fix it up afterward, as I have a few other points I wanted to discuss.

Reply:
Issuing HeadObject followed by GetObject opens the door to more eventual consistency with some object stores.

Reply:
yes, understand that but I don't see an alternative yet